src/sinks/util/sink.rs

```diff
@@ -0,0 +1,1193 @@
+
+                    trace!("Service ready; Sending batch.");
+                    let batch = self.batch.fresh_replace();
+
+                    let batch_size = batch.num_items();
+                    let request = batch.finish();
+
+                    let fut = self.service.call(request, batch_size);
+
+                    self.exec.spawn(fut).unwrap();
+
+                    // Disable linger timeout
+                    self.linger.take();
+                } else {
+                    // We have a batch but we can't send any items
+                    // most likely because we have not hit either
+                    // our batch size or the timeout. Here we want
+                    // to poll the inner futures but still return
+                    // NotReady if the linger timeout is not complete yet.
+                    if let Some(linger) = &mut self.linger {
+                        trace!("polling batch linger.");
+                        self.service.poll_complete()?;
+                        return linger.poll().map_err(Into::into);
+                    } else {
+                        return self.service.poll_complete();
+                    }
+                }
+            }
+        }
+    }
+
+    fn close(&mut self) -> Poll<(), Self::SinkError> {
+        trace!("closing batch sink.");
+        self.closing = true;
+        self.poll_complete()
+    }
+}
+
+impl<S, B, Request> fmt::Debug for BatchSink<S, B, Request>
+where
+    S: fmt::Debug,
+    B: fmt::Debug,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("BatchSink")
+            .field("service", &self.service)
+            .field("batch", &self.batch)
+            .field("settings", &self.settings)
+            .finish()
+    }
+}
+
+// === PartitionBatchSink ===
+
+type LingerDelay<K> = Box<dyn Future<Item = LingerState<K>, Error = ()> + Send + 'static>;
+
+/// A partition based batcher, given some `Service` and `Batch` where the
+/// input is partitionable via the `Partition` trait, it will hold many
+/// in flight batches.
+///
+/// This type is similar to `BatchSink` with the added benefit that it has
+/// more fine grained partitioning ability. It will hold many different batches
+/// of events and contain linger timeouts for each.
+///
+/// # Acking
+///
+/// Service based acking will only ack events when all prior request
+/// batches have been acked. This means if sequential requests r1, r2,
+/// and r3 are dispatched and r2 and r3 complete, all events contained
+/// in all requests will not be acked until r1 has completed.
+pub struct PartitionBatchSink<B, S, K, Request, E = DefaultExecutor> {
+    batch: B,
+    service: ServiceSink<S, Request>,
+    exec: E,
+    partitions: HashMap<K, B>,
+    settings: BatchSettings,
+    closing: bool,
+    sending: VecDeque<B>,
+    lingers: FuturesUnordered<LingerDelay<K>>,
+    linger_handles: HashMap<K, oneshot::Sender<K>>,
+}
```

Comment:
Is there not a way for us to make this a simple-ish wrapper/router around a collection of `BatchSink`s? Right now it feels like there's a decent amount of duplicate logic.

Reply:
I decided to not change this logic as much as possible. I think if we are going to even rewrite this we should just consider doing it in async/await imo.